File: packages/client-kotlin/compose/src/main/kotlin/ai/openclaw/spritecore/client/compose/CharacterAvatar.kt
package ai.openclaw.spritecore.client.compose

import ai.openclaw.spritecore.client.AgentAvatarSource
import ai.openclaw.spritecore.client.AnimationGraph
import ai.openclaw.spritecore.client.CharacterManifestEnvelope
import ai.openclaw.spritecore.client.CharacterManifestJson
import ai.openclaw.spritecore.client.SpriteAnimationPlayer
import ai.openclaw.spritecore.client.android.BitmapFrameSource
import android.graphics.Bitmap
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale

/**
 * Compose-driven avatar that renders a [CharacterManifestEnvelope] through the
 * SpriteCore animation engine. One code path covers sprite, atlas, and flat
 * states: every shape projects through [AnimationGraph.fromManifest] and
 * [SpriteAnimationPlayer], with [BitmapFrameSource] resolving the bytes.
 *
 * Pass a [bitmapTransform] to apply a per-frame crop or filter — e.g. a
 * watch face passing a top-half-square crop to extract a headshot from a
 * full-body sprite. Defaults to identity, so phone-sized avatars render the
 * frame as-authored.
 *
 * State-change inputs ([markerSignal] vs [currentState] / [playCount]):
 * - [markerSignal] takes precedence when non-null. Its monotonic
 *   [AgentAvatarSource.AvatarMarkerSignal.version] keys the launched effect
 *   so a model emitting the same `<<<wink-1>>>` twice in a row visibly
 *   replays — repeat markers don't get swallowed because `(state, count)`
 *   match. Pair with [AgentAvatarSource.setAgentState] on the producer side.
 * - When [markerSignal] is null, falls back to [currentState] +
 *   [playCount] keyed directly. Suitable for callers (e.g. a watch dial)
 *   that don't need same-marker replay semantics.
 *
 * Returns silently (paints nothing) when the manifest has no mode this
 * caller can render or when the first frame hasn't decoded yet; the caller's
 * fallback owns the empty state.
 */
@Composable
fun CharacterAvatar(
    agentId: String,
    envelope: CharacterManifestEnvelope,
    assetBytes: Map<String, ByteArray>,
    currentState: String?,
    contentDescription: String?,
    modifier: Modifier = Modifier,
    contentScale: ContentScale = ContentScale.Crop,
    bitmapTransform: (Bitmap) -> Bitmap = { it },
    playCount: Int? = null,
    markerSignal: AgentAvatarSource.AvatarMarkerSignal? = null,
) {
    val mode = remember(envelope.revision, agentId) {
        CharacterManifestJson.pickMode(envelope.manifest)
    }
    if (mode == null) return
    val graph = remember(envelope.revision, agentId, mode) {
        runCatching { AnimationGraph.fromManifest(envelope.manifest, mode) }.getOrNull()
    } ?: return

    // Key the FrameSource by revision so byte changes (new manifest version)
    // force re-decode of bitmaps; within a revision the sliceCache + decode
    // cache stay warm across recompositions.
    val frameSource = remember(envelope.revision, agentId, assetBytes) {
        BitmapFrameSource(assetBytes)
    }
    val player = remember(envelope.revision, agentId, mode) {
        SpriteAnimationPlayer(graph)
    }
    DisposableEffect(player) { onDispose { player.dispose() } }

    if (markerSignal != null) {
        LaunchedEffect(player, markerSignal.version) {
            val resolved = envelope.manifest.stateMap[markerSignal.state] ?: markerSignal.state
            if (envelope.manifest.content[mode]?.animations?.containsKey(resolved) == true) {
                player.requestState(resolved, playCount = markerSignal.count)
            }
        }
    } else {
        LaunchedEffect(player, currentState, playCount) {
            currentState?.takeIf { it.isNotBlank() }?.let { stateName ->
                val resolved = envelope.manifest.stateMap[stateName] ?: stateName
                if (envelope.manifest.content[mode]?.animations?.containsKey(resolved) == true) {
                    player.requestState(resolved, playCount)
                }
            }
        }
    }

    val ref by player.currentRef.collectAsState()
    val bitmap: Bitmap? = ref?.let { frameSource.frame(it) }
    if (bitmap == null) return
    val transformed = remember(bitmap, bitmapTransform) { bitmapTransform(bitmap) }
    Image(
        bitmap = transformed.asImageBitmap(),
        contentDescription = contentDescription,
        contentScale = contentScale,
        modifier = modifier.fillMaxSize(),
    )
}

File: packages/client-js/src/character-avatar.ts
import type { FrameRef, NodeGetCharacterManifestResult } from "./schema.js";
import { AnimationGraph } from "./animation-graph.js";
import { SpriteAnimationPlayer } from "./sprite-player.js";
import { ImageBitmapFrameSource, type AtlasFrame } from "./image-bitmap-frame-source.js";

/**
 * Per-frame transform applied before draw. The default is identity; pass a
 * function that returns a modified `AtlasFrame` (e.g. a different crop rect)
 * to add platform-specific framing — analogous to the Kotlin Compose
 * `bitmapTransform` and the SwiftUI `cgImageTransform` hooks.
 */
export type AvatarFrameTransform = (frame: AtlasFrame) => AtlasFrame;

export type CharacterAvatarOptions = {
  agentId: string;
  envelope: NodeGetCharacterManifestResult;
  assetBytes: Readonly<Record<string, Uint8Array>>;
  /** Initial state to request on mount. Falls back to the graph's default. */
  initialState?: string | null;
  /** Initial play count (mirrors `<<<state-N>>>`). Optional. */
  initialPlayCount?: number | null;
  /** Frame-time transform; defaults to identity. */
  transform?: AvatarFrameTransform;
};

export type CharacterAvatarController = {
  /** Request a state change; mirrors the Compose `currentState` parameter. */
  setState(state: string | null, playCount?: number | null): void;
  /** Tear down the player, frame source, and canvas subscription. */
  dispose(): Promise<void>;
};

/**
 * Mount the SpriteCore avatar engine onto an `HTMLCanvasElement`. Mirrors the
 * Kotlin Compose `CharacterAvatar` and the SwiftUI `CharacterAvatarView`:
 * one code path covers sprite, atlas, and flat states via
 * `AnimationGraph.fromManifest` + `SpriteAnimationPlayer`, with
 * `ImageBitmapFrameSource` resolving the bytes.
 *
 * Drawing fits the frame into the canvas's intrinsic `width`/`height` (i.e.
 * the backing-store resolution, not CSS size). Set those on the canvas to
 * the desired pixel dimensions before mounting; CSS handles layout.
 *
 * Returns a controller with `setState` and `dispose`. If the manifest has no
 * mode this caller can render, returns a no-op controller — the caller's
 * fallback owns the empty state.
 */
export function mountCharacterAvatar(
  canvas: HTMLCanvasElement,
  opts: CharacterAvatarOptions,
): CharacterAvatarController {
  const { envelope, assetBytes, initialState, initialPlayCount, transform } = opts;
  const mode = pickMode(envelope.manifest);
  if (mode === null) {
    return noopController();
  }

  let graph: AnimationGraph;
  try {
    graph = AnimationGraph.fromManifest(envelope.manifest, mode);
  } catch {
    return noopController();
  }

  const ctx = canvas.getContext("2d");
  if (!ctx) {
    return noopController();
  }

  const frameSource = new ImageBitmapFrameSource(assetBytes);
  const player = new SpriteAnimationPlayer(graph);
  let disposed = false;

  // Kick decode of every ref so frames are ready by the time the player
  // emits them. Once primed, `frame()` resolves synchronously.
  void frameSource.prefetchAll();

  const draw = (ref: FrameRef | null): void => {
    if (disposed) return;
    if (!ref) {
      ctx.clearRect(0, 0, canvas.width, canvas.height);
      return;
    }
    const sliced = frameSource.frame(ref);
    if (!sliced) {
      // Frame not yet decoded — kick prefetch and bail. The player keeps
      // emitting at frame rate, so the next ref change picks up the decode.
      void frameSource.prefetch(ref.ref);
      return;
    }
    const t = transform ? transform(sliced) : sliced;
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    ctx.drawImage(t.image, t.sx, t.sy, t.sw, t.sh, 0, 0, canvas.width, canvas.height);
  };

  const unsubscribe = player.currentRef.subscribe(draw);

  const requestStateInternal = (state: string | null, playCount: number | null): void => {
    if (disposed || !state) return;
    const resolved = envelope.manifest.stateMap[state] ?? state;
    if (envelope.manifest.content[mode]?.animations[resolved] === undefined) return;
    void player.requestState(resolved, playCount);
  };

  if (initialState !== undefined && initialState !== null) {
    requestStateInternal(initialState, initialPlayCount ?? null);
  }

  return {
    setState(state, playCount = null) {
      requestStateInternal(state, playCount);
    },
    async dispose() {
      if (disposed) return;
      disposed = true;
      unsubscribe();
      await player.dispose();
      frameSource.dispose();
    },
  };
}

/** Pick the first mode in `manifest.modes` whose content is present. */
function pickMode(manifest: NodeGetCharacterManifestResult["manifest"]): string | null {
  for (const m of manifest.modes) {
    if (m in manifest.content) return m;
  }
  return null;
}

function noopController(): CharacterAvatarController {
  return {
    setState() {},
    async dispose() {},
  };
}
